src/core/QCheck2.ml

```diff
@@ -0,0 +1,1868 @@
+      towards the destination. *)
+  let float_towards destination x =
+    Float.(number_towards ~equal ~div ~add ~sub ~of_int) ~destination x |> Seq.take 15
+
+  let int_aggressive_towards (destination : int) (n : int) : int Seq.t =
+    Seq.unfold (fun current ->
+        if current = n then None
+        else if current < n then let next = succ current in Some (next, next)
+        else let next = pred current in Some (next, next)
+      ) destination
+
+  let int_aggressive n = int_aggressive_towards 0 n
+
+end
+
+module Tree = struct
+  type 'a t = Tree of 'a * ('a t) Seq.t
+
+  let root (Tree (root, _) : 'a t) : 'a = root
+
+  let children (Tree (_, children) : 'a t) : ('a t) Seq.t = children
+
+  let rec pp ?(depth : int option) (inner_pp : Format.formatter -> 'a -> unit) (ppf : Format.formatter) (t : 'a t) : unit =
+    let open Format in
+    let Tree (x, xs) = t in
+    let wrapper_box inner =
+      pp_open_vbox ppf 2;
+      pp_print_string ppf "Tree(";
+      pp_print_break ppf 0 0;
+      inner ();
+      pp_close_box ppf ();
+      pp_print_break ppf 0 0;
+      pp_print_string ppf ")"
+    in
+    let inner () =
+      pp_open_vbox ppf 2;
+      pp_print_string ppf "Node(";
+      pp_print_break ppf 0 0;
+      inner_pp ppf x;
+      pp_close_box ppf ();
+      pp_print_break ppf 0 0;
+      pp_print_string ppf "),";
+      pp_print_break ppf 1 0;
+      pp_open_vbox ppf 2;
+      pp_print_string ppf "Shrinks(";
+
+      if Option.fold depth ~none:false ~some:(fun depth -> depth <= 0) then (
+        pp_print_string ppf "<max depth reached>";
+        pp_close_box ppf ())
+      else if Seq.is_empty xs then pp_close_box ppf ()
+      else
+        (pp_print_break ppf 0 0;
+         pp_print_list
+           ~pp_sep:(fun ppf () -> pp_print_string ppf ","; pp_print_space ppf ())
+           (pp ?depth:(Option.map pred depth) inner_pp)
+           ppf
+           (List.of_seq xs);
+         pp_close_box ppf ();
+         pp_print_break ppf 0 0);
+
+      pp_print_string ppf ")"
+
+    in
+    wrapper_box inner
+
+  let rec map (f : 'a -> 'b) (a : 'a t) : 'b t =
+    let Tree (x, xs) = a in
+    let y = f x in
+    let ys = Seq.map (fun smaller_x -> map f smaller_x) xs in
+    Tree (y, ys)
+
+  (** Note that parameter order is reversed. *)
+  let (>|=) a f = map f a
+
+  let rec ap (f : ('a -> 'b) t) (a : 'a t) : 'b t =
+    let Tree (x0, xs) = a in
+    let Tree (f0, fs) = f in
+    let y = f0 x0 in
+    let ys = Seq.append (Seq.map (fun f' -> ap f' a) fs) (Seq.map (fun x' -> ap f x') xs) in
+    Tree (y, ys)
+
+  let (<*>) = ap
+
+  let liftA2 (f : 'a -> 'b -> 'c) (a : 'a t) (b : 'b t) : 'c t =
+    (a >|= f) <*> b
+
+  let rec bind (a : 'a t) (f : 'a -> 'b t) : 'b t =
+    let Tree (x, xs) = a in
+    let Tree (y, ys_of_x) = f x in
+    let ys_of_xs = Seq.map (fun smaller_x -> bind smaller_x f) xs in
+    let ys = Seq.append ys_of_xs ys_of_x in
+    Tree (y, ys)
+
+  let (>>=) = bind
+
+  let pure x = Tree (x, Seq.empty)
+
```

Comment:
That's an excellent idea!

Actually I would like to push it further if you are ok with it:
- provide `Tree.make_unfold` as you suggest (though if you don't mind, I think it's better to name it `unfold`)
- **remove** `Gen.make_primitive` (I hated it because it exposes the underlying Random library, which would force a breaking change if the Random lib is replaced by another one tomorrow, e.g. to get random splitting)
- add `Tree.add_shrinks : ('a -> 'a Seq.t) -> 'a t -> 'a t` in the same spirit as https://hackage.haskell.org/package/hedgehog-1.0.5/docs/Hedgehog-Gen.html#v:shrink
- add `Tree.prune : 'a t -> 'a t` that removes all shrinks, similar to https://hackage.haskell.org/package/hedgehog-1.0.5/docs/Hedgehog-Gen.html#v:prune

API-wise this means the "underlying" implementation we provide to users is `Tree` to manually craft generators that fit their needs:
- users can reuse an existing generator and change its shrinks only, using `prune` and `add_shrinks`
- since we provide generators for all primitive types of `Random`, it's ok to NOT expose any `Random` API, as users can compose a primitive generator

We may want additionally to provide some equivalent functions at `Gen` level to lower the needed code/noise for users, and round up the API experience

Does that sound good? I don't know if I'm clear enough 😅

Reply:
Moving this to https://github.com/c-cube/qcheck/issues/122